src/trace/log_backend_probe.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019,2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/zephyr.h>
+
+#include <sof/audio/buffer.h>
+#include <sof/ipc/topology.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/probe/probe.h>
+
+/*
+ * A lock is needed as log_process() and log_panic() have no internal locks
+ * to prevent concurrency. Meaning if log_process is called after
+ * log_panic was called previously, log_process may happen from another
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock lock;
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_ADSP_OUTPUT_DEFAULT;
+
+/**
+ * 80 bytes seems to catch most sensibly sized log message lines
+ * in one go letting the trace out call output whole complete
+ * lines. This avoids the overhead of a spin lock in the trace_out
+ * more often as well as avoiding entwined characters from printk if
+ * LOG_PRINTK=n.
+ */
+#define LOG_BUF_SIZE 80
+static uint8_t log_buf[LOG_BUF_SIZE];
+
+static struct comp_buffer *buffer;
+
+struct ipc_comp_dev logging_comp;
+
+static int char_out(uint8_t *data, size_t length, void *ctx)
+{
+	/* allocate new buffer */
+	if (!buffer) {
+		buffer = buffer_alloc(LOG_BUF_SIZE, 0, 128);
+		if (!buffer) {
+			/* LOG_ERR("probe log buffer alloc failed"); */
+			return 0;
+		}
+
+		buffer->id = PROBE_LOGGING_BUFFER_ID_MAGIC;
+
+		logging_comp.type = COMP_TYPE_BUFFER;
+		logging_comp.cb = buffer;
+	}
+
+	memcpy(buffer->stream.addr, data, length);
+
+	comp_update_buffer_produce(buffer, length);
+
+	return length;
+}
+
+LOG_OUTPUT_DEFINE(log_output_adsp_probe, char_out, log_buf, sizeof(log_buf));
+
+static uint32_t format_flags(void)
+{
+	uint32_t flags = LOG_OUTPUT_FLAG_LEVEL | LOG_OUTPUT_FLAG_TIMESTAMP;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP)) {
+		flags |= LOG_OUTPUT_FLAG_FORMAT_TIMESTAMP;
+	}
+
+	return flags;
+}
+
+#undef panic
```

Comment:
Ugh, SOF headers redefine panic(). Took a while to figure out why code lifted from Zephyr to SOF simply did not compile anymore.

Reply:
So this will _only_ work with zephyr, no xtos support?

Reply:
> So this will _only_ work with zephyr, no xtos support?

Yes. No point doing this for xtos.